aws/rust-runtime/aws-types/src/profile/source.rs

```diff
@@ -0,0 +1,312 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+use crate::environment;
+use std::borrow::Cow;
+use std::io::ErrorKind;
+use std::path::{Component, Path, PathBuf};
+
+/// In-memory source of profile data
+pub struct Source {
+    /// Contents and path of ~/.aws/config
+    pub config_file: File,
+
+    /// Contents and path of ~/.aws/credentials
+    pub credentials_file: File,
+
+    /// Profile to use
+    ///
+    /// Overridden via `$AWS_PROFILE`, defaults to `default`
+    pub profile: Cow<'static, str>,
+}
+
+/// In-memory configuration file
+pub struct File {
+    pub path: String,
+    pub contents: String,
+}
+
+/// Load a [Source](Source) from a given environment and filesystem.
+pub fn load(proc_env: &environment::ProcessEnvironment, fs: &environment::Fs) -> Source {
+    let config = tracing::info_span!("load_config_file")
+        .in_scope(|| read(&fs, &proc_env, "~/.aws/config", "AWS_CONFIG_FILE"));
```

Comment:
Wasn't obvious that `read` was replacing `~` on different OSes. Should we just pass in the `.aws/config` part and just always look in the user directory for it? Or have a lower-level read function that takes a full path, and a higher level one that resolves it?

Better yet, should we have an enum to represent the `config` and `credentials` files with a `path()` function?

Reply:
there's some wrinkles here that we also need to resolve the home directory from what the user might give us in an environment variable

Reply:
so having implicit home directory searching is probably going to create more complexity (since we still need generic home directory replacement)

Reply:
but I can split up the path resolution and the home directory resolution for clarity.